craft_providers/lxd/lxd_instance.py

```diff
@@ -0,0 +1,494 @@
+# it under the terms of the GNU General Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""LXD Instance Executor."""
+
+import io
+import logging
+import os
+import pathlib
+import shutil
+import subprocess
+import tempfile
+from typing import Any, Dict, List, Optional
+
+from craft_providers import errors
+from craft_providers.util import env_cmd
+
+from .. import Executor
+from .errors import LXDError
+from .lxc import LXC
+
+logger = logging.getLogger(__name__)
+
+
+def _rootify_lxc_command(
+    command: List[str],
+    *,
+    env: Optional[Dict[str, Optional[str]]] = None,
+) -> List[str]:
+    """Wrap a command to run as root with specified environment.
+
+    Formulate command to:
+
+        - Use sudo to run as root (LXD defaults to root, but doesn't initialize
+          environment like sudo will, e.g. read /etc/environment).  For
+          consistency with Multipass and other providers, always use sudo.
+
+        - Configure sudo to set home directory.
+
+        - Account for environment flags in env, if any.
+
+    :param command: Command to execute.
+    :param env: Additional environment flags to set.
+
+    :returns: List of command strings for multipass exec.
+    """
+    sudo_cmd = ["sudo", "-H", "--"]
+
+    if env is not None:
+        sudo_cmd += env_cmd.formulate_command(env)
+
+    return sudo_cmd + command
+
+
+class LXDInstance(Executor):
+    """LXD Instance Lifecycle."""
+
+    def __init__(
+        self,
+        *,
+        name: str,
+        project: str = "default",
+        remote: str = "local",
+        lxc: Optional[LXC] = None,
+    ):
+        super().__init__()
+
+        self.name = name
+        self.project = project
+        self.remote = remote
+
+        if lxc is None:
+            self.lxc = LXC()
+        else:
+            self.lxc = lxc
+
+    def create_file(
+        self,
+        *,
+        destination: pathlib.Path,
+        content: io.BytesIO,
+        file_mode: str,
+        group: str = "root",
+        user: str = "root",
+    ) -> None:
+        """Create file with content and file mode.
+
+        :param destination: Path to file.
+        :param content: Contents of file.
+        :param file_mode: File mode string (e.g. '0644').
+        :param group: File group owner/id.
+        :param user: File user owner/id.
+
+        :raises LXDError: On unexpected error.
+        """
+        with tempfile.NamedTemporaryFile(delete=False) as temp_file:
+            shutil.copyfileobj(content, temp_file)
+            temp_file.flush()
```

Comment:
why this flush is necessary? the temp file is closed when the context manager exits, so...

Reply:
I think I prior version of this code used it before exiting.  I can't recall a specific problem so I'm removing it (and the integration tests don't show any issue).

Reply:
Fixed